src/libstore/build/derivation-goal.cc

```diff
@@ -1322,7 +1320,14 @@ std::pair<bool, DrvOutputs> DerivationGoal::checkPathValidity()
     if (!drv->type().isPure()) return { false, {} };
 
     bool checkHash = buildMode == bmRepair;
-    auto wantedOutputsLeft = wantedOutputs;
+    auto wantedOutputsLeft = std::visit(overloaded {
+        [&](const OutputsSpec::All &) {
+            return StringSet {};
+        },
+        [&](const OutputsSpec::Names & names) {
+            return names;
+        },
+    }, wantedOutputs.raw());
```

Comment:
This  is very similar to the code above (`realWantedOutputs`), so maybe this could be merged into a method on `OutputSpec`, e.g. `outputSpec.names_or(...)`.